UPGRADE-GUIDE.md

```diff
@@ -0,0 +1,159 @@
+# Rack 3 Upgrade Guide
+
+This document is a work in progress, but outlines some of the key changes in Rack 3 which you should be aware of in order to update your server, middleware and/or applications.
+
+## Request Changes
+
+### `rack.version` is no longer required
+
+Previously, the "rack protocol version" was available in `rack.version` but it was not practically useful, so it has been removed as a requirement.
+
+### `rack.multithread`/`rack.multiprocess`/`rack.run_once` are no longer required
+
+Previously, servers tried to provide these keys to reflect the execution environment. These come too late to be useful, so they have been removed as  a requirement.
+
+### `rack.hijack?` now only applies to partial hijack
+
+Previously, both full and partial hijiack were controlled by the presence and value of `rack.hijack?`. Now, it only applies to partial hijack (which itself has been effectively replaced by streaming bodies).
+
+### `rack.hijack` alone indicates that you can execute a full hijack
+
+Previously, `rack.hijack?` had to be truthy, as well as having `rack.hijack` present in the request environment. Now, the presence of the `rack.hijack` callback is enough.
+
+### `rack.hijack_io` is removed
+
+Previously, the server would try to set `rack.hijack_io` into the request environment when `rack.hijack` was invoked for a full hijack. This was often impossible if a middleware had called `env.dup`, so this requirement has been dropped entirely.
+
+### `rack.input` is no longer required to be rewindable
+
+Previosuly, `rack.input` was required to be rewindable, i.e. `io.seek(0)` but this was only generally possible with a file based backing, which prevented efficient streaming of request bodies. Now, `rack.input` is not required to be rewindable.
+
+## Response Changes
+
+### Response must be mutable
+
+Rack 3 requires the response Array `[status, headers, body]` to be mutable. Existing code that uses a frozen response will need to be changed:
+
+```ruby
+NOT_FOUND = [404, {}, ["Not Found"]].freeze
+
+def call(env)
+  ...
+	return NOT_FOUND
```

Comment:
Please remove all use of tabs in this file.  Ruby code should be indented using 2 spaces.

Reply:
Thanks for pointing that out, we should consider adding a `.editorconfig`.